pkg/sessions/redis/client.go

```diff
@@ -18,16 +22,54 @@ var _ Client = (*client)(nil)
 
 type client struct {
 	*redis.Client
+	locker *redislock.Client
+	locks  map[string]*redislock.Lock
 }
 
 func newClient(c *redis.Client) Client {
-	return &client{Client: c}
+	return &client{
+		Client: c,
+		locker: redislock.New(c),
+		locks:  map[string]*redislock.Lock{},
+	}
 }
 
 func (c *client) Get(ctx context.Context, key string) ([]byte, error) {
+	if c.locks[key] != nil {
+		for {
+			ttl, err := c.locks[key].TTL(ctx)
+			if err != nil {
+				return nil, err
+			}
+			if ttl <= 0 {
+				break
+			}
+		}
+	}
 	return c.Client.Get(ctx, key).Bytes()
 }
 
+func (c *client) Lock(ctx context.Context, key string, expiration time.Duration) error {
```

Comment:
It looks like `redislock` doesn't have a way to `Peek` at whether a lock is already taken without trying to obtain it ourselves? I really only see `Obtain` exposed.

That functionality would be nice for us to block and wait for a lock to be done without trying to take it (so we can then pull a fresh session from the store).

Reply:
I want to avoid a long queue of requests waiting for the session refresh lock to be done - only to have to all lock/unlock themselves to pull the fresh session.

Reply:
You are right, unfortunately, they are not supporting that right now.

Reply:
I solved it with just doing retries with Load, if LoadWithLock failed once. What do you think. This is not yet tested, just an idea.

Reply:
> doesn't have a way to Peek at whether a lock is already taken

Is this something we could implement on top of it possibly?

> I want to avoid a long queue of requests waiting for the session refresh lock to be done - only to have to all lock/unlock themselves to pull the fresh session.

Lock/unlock should be pretty quick once we know that the session has been revalidated right? What I don't want to do is overcomplicate the original implementation by having strict requirements that may not be 100% necessary. This PR is quite large, I wonder if we can break it into steps and get a minimal implementation that may not be polished, and iterate to make it more efficient?

Reply:
I'm gonna try to spike out an implementation of the proactive refreshing that is tied to this -- that being separate might make us able to fully focus on the locking functionality in this PR.

Reply:
@NickMeves: Great, let me know, when this is done and I can continue here.

Reply:
Draft is up here: https://github.com/oauth2-proxy/oauth2-proxy/pull/1086

Once I get some feedback on the design choices from @JoelSpeed I'll wrap it up. It's admittedly raw since I did it from my couch while watching TV 😅